rfcs/0128-selective-auto-merge-on-bot-upgrades.md

```diff
@@ -0,0 +1,53 @@
+---
+feature: Selective auto-merge on bot upgrades
+start-date: 2022-07-07
+author: superherointj
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a `meta.autoMerge` attribute to packages to allow committers to delegate merge rights to package maintainers for bot auto updates.
+
+# Motivation
+[motivation]: #motivation
+
+* Reduce pending PRs for review and merge in nixpkgs.
+* Save reviewers and commiters time.
+* Speed up package upgrades.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new `meta.autoMerge` package attribute with type `bool` defaulting to `false`. To be documented in Nixpkgs manual.
+
+OfBorg reads `meta.autoMerge` value and sets `13.auto-merge` label to PR.
+
+GitHub Actions will trigger a merge when the following 4 conditions are met:
+1) `@r-ryantm` is PR's author.
+2) Has `13.auto-merge` label.
+3) All CI checks passed.
+4) A package maintainer has approved PR. (Label: `12.approved-by: package-maintainer`)
+
+Due maintainers being able to add labels in GitHub, an extra check for labels might be necessary in GitHub Actions before merge can happen.
+
+The appropriateness of setting `meta.autoMerge` is left up to committers.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Some possible security issue.
```

Comment:
We need to discuss the drawback about having this can reduce the trustfulness of the nixpkgs as a whole. For example, what about `node-packages` updates that changes a huge file?

Even if this is not affected by this change, people seeing PRs being merged automatically can see this as a possibly security hazard.

Reply:
All PRs to be auto-merged are still going to be reviewed by humans, maintainers in the case. And commiters have to enable autoMerge per package, considering all the drawbacks it might cause. In case a package is not fit for this feature, the expectation is it won't be enabled. The aim is to add a capability and the judgement is to be left to humans.

Reply:
> For example, what about `node-packages` updates that changes a huge file?

nodePackages already receives next to no review because it is almost impossible with sometimes over 5000 lines diff.

> All PRs to be auto-merged are still going to be reviewed by humans, maintainers in the case.

The entrance barrier to be a maintainer is very low and so far didn't include any necessary review tasks. I don't think we can assume that there will be a human review in every case. In the end no one stops people from just approving their PRs without testing them or even fully automate it.

When interacting with committers you build trust over time especially if they are quite active and you see PRs/review/comments from them weekly or even daily. If a maintainer is quite active they usually are promoted to committers at some point and in that time they also gathered quite a bit of trust. I am not sure where the average maintainer fits in here but I personally would think they have less overhaul trust compared to committers.